post/2023/beyond-sf/index.qmd

```diff
@@ -0,0 +1,305 @@
+
+:::{.callout-note}
+Read more about **sf** in [Spatial Data Science](https://r-spatial.org/book/07-Introsf.html#sec-sfintro)
+:::
+
+But `sf` needs to be understood in a more fundamental way. `sf` is named after the [Simple Feature Access Standard](https://en.wikipedia.org/wiki/Simple_Features). Simple features are an agreed upon way to represent "geometric primitives"—things like `Point`s, `LineString`s, `Polygon`s, and their `Multi`- types. The `sf` package builds a hierarchy off of these. We have `sfg`, `sfc`, and `sf` objects.
+
+[![](https://r-spatial.github.io/sf/articles/sf_xfig.png)](https://r-spatial.github.io/sf/articles/sf1.html#sf-objects-with-simple-features)
+
+### `sfg` objects
+
+At the core is the `sfg` class which are representations of simple feature geometries. `sfg` class objects are representations of a simple feature. They are a single geometry at a time. We can think of them as a scalar value (even though R does not have the concept of a scalar).
+
+```{r}
+library(sf)
+
+# create a point
+pnt <- st_point(c(0, 10))
+
+# create a line
+ln <- st_linestring(matrix(c(0, 1, 0, 0), ncol = 2))
+
+class(pnt)
+class(ln)
+```
+
+Each scalar value can be used independently which is useful in itself. 
+
+```{r}
+st_length(ln)
+```
+`sfg` objects are very simple objects constructed of numeric vectors, matrices, and lists of matrices. They also have no sense of a coordinate reference system (CRS). More often than not, though, we want to have a vector of geometries—one for each element in a dataset, for example. In sf, a vector of geometries is stored in an `sfc` object.
+
+### `sfc` objects
+
+`sfc` is short for **s**imple **f**eature **c**olumn. You might think that you could create a vector of geometries by combining them using `c()` but that would be wrong
+
+```{r}
+c(pnt, pnt)
+c(pnt, ln)
+```
+`sfg` objects behave like scalars so combining them creates either a `Multi`- type of the `sfg` or a geometry collection (another type of simple feature).
+
+To create a vector of geometries you must use `st_sfc()`. `st_sfc()` is the construct for creating a "simple feature geometry list column." It lets you also set a number of attributes that are associated with the vector. 
+
+```{r}
+st_sfc(pnt, pnt)
+```
+
+Each `sfc` object contains attributes such as a CRS (optional), bounding box, and precision.
+
+```{r}
+c(
+  st_sfc(pnt, pnt), 
+  st_sfc(pnt),
+  st_sfc(ln)
+)
+```
+
+Under the hood, this is just a list of `sfg` objects. Remember, it's not magic! 
+
+```{r}
+unclass(st_sfc(pnt, pnt))
+```
+
+Since `sfc` objects are vectors, they can be included as a column in a data frame. 
+
+```{r}
+df <- data.frame(
+  geo = st_sfc(pnt, pnt)
+)
+
+df
+
+class(df)
+```
+
+## Geometry and data.frames
+
+Having geometry included in a data frame was a huge win for the R community because this means that attributes can be included along with geometries. Further, geometries in a data frame means that they are dplyr compatible. 
```

Comment:
```suggestion
Having geometry included in a data frame was a huge win for the R community because this means that attributes can be included along with geometries. Further, geometries in a data frame means that they are **dplyr** compatible. 
```